### ISSUE TITLE
playlist m3u8 sometimes isn't immediately available (HTTP 404) after a long space just ended

### ISSUE THREAD

**Issue by fireattack** [HoloArchivists/tslazer#5]
This is what I tried to solve with #4 but I noticed the cause is actually this:

When a long Space just ended, the playlist_xxx given in master_playlist *sometimes* won't actually be available (HTTP 404) for quite awhile. 

Here are the logs from a [monkey-patched version](https://github.com/fireattack/tslazer) (based on V1) I have, just to demonstrate the issue:

```
D:\temp\tslazer>tslazer.py -s 1kvKpmXgMbMGE
[DEBUG] Update playlist URLs...
Space Found!
Space Title: 絵を描くスペース
Space Host Username: hoshiume0404
Space Host Display Name: 干し梅
Space Master URL:
https://prod-fastly-ap-northeast-1.video.pscp.tv/Transcoding/v1/hls/pxBnReI2LcTHLle-LVyb2qcazCAsPsAz1RjTGdm1WjIP3-dduSt965Bh7gp2JvjfTpheVRcPOFOaZeDnsKEXTw/transcode/ap-northeast-1/periscope-replay-direct-prod-ap-northeast-1-public/eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCIsInZlcnNpb24iOiIyIn0.eyJIZWlnaHQiOjEwMCwiS2JwcyI6MzUwLCJUcmFuc2NvZGVBdWRpbyI6dHJ1ZSwiV2lkdGgiOjEwMH0.xKaxg744Z6K1TwhoGXRWJ3A3FlSeNBKxxUEyorWq8qY/audio-space/playlist_16760973110719389214.m3u8
Space Dynamic URL:
https://prod-fastly-ap-northeast-1.video.pscp.tv/Transcoding/v1/hls/pxBnReI2LcTHLle-LVyb2qcazCAsPsAz1RjTGdm1WjIP3-dduSt965Bh7gp2JvjfTpheVRcPOFOaZeDnsKEXTw/non_transcode/ap-northeast-1/periscope-replay-direct-prod-ap-northeast-1-public/audio-space/master_playlist.m3u8
Chat Token:
eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE2ODU4NTk3NjQsImFtYmlndW91c191c2VyX2lkIjoiZXlKVWQybDBkR1Z5VlhObGNrbGtJam94TmpnMU56Y3pNelkwTmpNeE9EWXlOelo5IiwiYnJvYWRjYXN0X2lkIjoiMWt2S3BtWGdNYk1HRSIsImxvd19sYXRlbmN5Ijp0cnVlLCJyZWFkX29ubHkiOmZhbHNlLCJwYXJ0aWNpcGFudF9pbmRleCI6NjgzOTg0ODQwfQ.fCzL8M6_X_1C9IeUV4jaKr9BHqpcoFSzWPbMo0PfElc
Downloading to 230603 @hoshiume0404 絵を描くスペース-twitter-space-1kvKpmXgMbMGE.m4a
Waiting for space to end...
Space Ended. Wait 1 minute for the recording to be processed.
[DEBUG] current master_url filename: playlist_16760973110719389214.m3u8
[DEBUG] Update playlist URLs...
[DEBUG] current master_url filename: playlist_16760972807730874215.m3u8
[DEBUG] current master_url:
https://prod-fastly-ap-northeast-1.video.pscp.tv/Transcoding/v1/hls/pxBnReI2LcTHLle-LVyb2qcazCAsPsAz1RjTGdm1WjIP3-dduSt965Bh7gp2JvjfTpheVRcPOFOaZeDnsKEXTw/transcode/ap-northeast-1/periscope-replay-direct-prod-ap-northeast-1-public/eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCIsInZlcnNpb24iOiIyIn0.eyJIZWlnaHQiOjEwMCwiS2JwcyI6MzUwLCJUcmFuc2NvZGVBdWRpbyI6dHJ1ZSwiV2lkdGgiOjEwMH0.xKaxg744Z6K1TwhoGXRWJ3A3FlSeNBKxxUEyorWq8qY/audio-space/playlist_16760972807730874215.m3u8
[DEBUG] request status code: 404
[DEBUG] failed to get playlist m3u8, retry after 10 seconds...
[DEBUG] Update playlist URLs...
[DEBUG] current master_url filename: playlist_16760970710781794128.m3u8
[DEBUG] current master_url:
https://prod-fastly-ap-northeast-1.video.pscp.tv/Transcoding/v1/hls/pxBnReI2LcTHLle-LVyb2qcazCAsPsAz1RjTGdm1WjIP3-dduSt965Bh7gp2JvjfTpheVRcPOFOaZeDnsKEXTw/transcode/ap-northeast-1/periscope-replay-direct-prod-ap-northeast-1-public/eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCIsInZlcnNpb24iOiIyIn0.eyJIZWlnaHQiOjEwMCwiS2JwcyI6MzUwLCJUcmFuc2NvZGVBdWRpbyI6dHJ1ZSwiV2lkdGgiOjEwMH0.xKaxg744Z6K1TwhoGXRWJ3A3FlSeNBKxxUEyorWq8qY/audio-space/playlist_16760970710781794128.m3u8
[DEBUG] request status code: 404
[DEBUG] failed to get playlist m3u8, retry after 10 seconds...
[DEBUG] Update playlist URLs...
[DEBUG] current master_url filename: playlist_16760971421756524434.m3u8
[DEBUG] current master_url:
https://prod-fastly-ap-northeast-1.video.pscp.tv/Transcoding/v1/hls/pxBnReI2LcTHLle-LVyb2qcazCAsPsAz1RjTGdm1WjIP3-dduSt965Bh7gp2JvjfTpheVRcPOFOaZeDnsKEXTw/transcode/ap-northeast-1/periscope-replay-direct-prod-ap-northeast-1-public/eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCIsInZlcnNpb24iOiIyIn0.eyJIZWlnaHQiOjEwMCwiS2JwcyI6MzUwLCJUcmFuc2NvZGVBdWRpbyI6dHJ1ZSwiV2lkdGgiOjEwMH0.xKaxg744Z6K1TwhoGXRWJ3A3FlSeNBKxxUEyorWq8qY/audio-space/playlist_16760971421756524434.m3u8
[DEBUG] request status code: 404
[DEBUG] failed to get playlist m3u8, retry after 10 seconds...
[DEBUG] Update playlist URLs...
[DEBUG] current master_url filename: playlist_16760968411484182689.m3u8
[DEBUG] current master_url:
https://prod-fastly-ap-northeast-1.video.pscp.tv/Transcoding/v1/hls/pxBnReI2LcTHLle-LVyb2qcazCAsPsAz1RjTGdm1WjIP3-dduSt965Bh7gp2JvjfTpheVRcPOFOaZeDnsKEXTw/transcode/ap-northeast-1/periscope-replay-direct-prod-ap-northeast-1-public/eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCIsInZlcnNpb24iOiIyIn0.eyJIZWlnaHQiOjEwMCwiS2JwcyI6MzUwLCJUcmFuc2NvZGVBdWRpbyI6dHJ1ZSwiV2lkdGgiOjEwMH0.xKaxg744Z6K1TwhoGXRWJ3A3FlSeNBKxxUEyorWq8qY/audio-space/playlist_16760968411484182689.m3u8
[DEBUG] request status code: 200
[DEBUG] get 5225 chunks.
5225/5225 chunks downloaded.
Finished Downloading Chunks.
[DEBUG] command is ffmpeg -loglevel error -stats -i "230603 @hoshiume0404 絵を描くスペース-twitter-space-1kvKpmXgMbMGE_merged.aac" -c copy -metadata title="絵を描くスペース" -metadata artist="hoshiume0404" "230603 @hoshiume0404 絵を描くスペース-twitter-space-1kvKpmXgMbMGE.m4a"
size=  181885kB time=04:21:16.35 bitrate=  95.0kbits/s speed=2.55e+03x
Successfully Downloaded Twitter Space 230603 @hoshiume0404 絵を描くスペース-twitter-space-1kvKpmXgMbMGE.m4a
```

I checked the `twitter-space-overhaul` branch, and I think the issue would be the same there. 

I don't have time to write a proper PR for overhaul version at the moment, but I think you get the idea about the nature of this bug. Maybe you or someone else can write a patch for this.

**Comment by ef1500** [comment]
Hello! I've run into this error on occasion as well, and I'm unsure whether or not it's on my end or twitter's. There's a possibility that changing the url from "master_playlist" back to "dynamic_playlist" might do the trick; because that's what's used when you listen from the app/web client. That would require a dynamic downloader, I'd assume, because the way that the dynamic playlists work is that they only give a certain number of chunks at a time rather than the whole playlist.

Alternatively, we could implement a timed retry like you have in your patched version. Either method would work, I think. 
I might just add both and make the retry method the default, depending on how much time I have in the coming weeks and days.

**Comment by ef1500** [comment]
It seems the issue lies with your last pull request.
While it *does* get the job accomplished, sometimes when it tries to get the latest playlists for an unarchived twitter space, it fails, because well, the space doesn't exist anymore.

If you already have the master playlist, you don't need to get it again.

**Comment by fireattack** [comment]
Just to clear it up. 

Firstly, what we saved in `self.playlist.master_url` isn't actually master m3u8 (`master_playlist.m3u8`), it's the "playlist_{timestamp}.m3u8" that is embedded in `master_playlist.m3u8` (a subplaylist). So that's a misnomer.

This playlist is a static, VOD (`#EXT-X-ENDLIST`) m3u8 which only has all the chunks the server saved **so far**, and it will NOT update.

So it itself is kinda useless. We NEED to update it (by re-fetching `master_playlist.m3u8`, which is saved as `self.playlist.dyn_url` in our code, another misnomer. But this one doesn't matter because they're interchangeable) when we're ready to download chunks. If we don't after the Space ended, we will only get chunks up to when we started to run `tslazer`, which obviously is wrong.

(And there is no guarantee this playlist is even HTTP 200 later: you can try the initial playlist I got from the above log, `playlist_16760973110719389214.m3u8` -- it's 404 now.)

And about my PR: in hindsight, is was actually redundant , because at

https://github.com/HoloArchivists/tslazer/blob/964abc2bc305c9c1ccacc3bbd0393a9f73c374eb/TwitterSpace.py#LL345C1-L347C97

```python
            self.playlists.master_url = TwitterSpace.getPlaylists(dyn_url=self.playlists.dyn_url).master_url
            livechunks = TwitterSpace.getChunks(self.playlists)
            TwitterSpace.downloadChunks(livechunks, self.filenameformat, self.path, m4aMetadata)
```

We already tried to update playlist before `getChunks()`. So what I added in last PR is just trying it one more time immediately (inside of `getChunks()`), which has virtually no effect.

All these are irrelevant to the issue of playlist_{timestamp} being 404 -- this is on Twitter's sever's end.

Essentially, Twitter constantly update `master_playlist.m3u8` to direct to the newest `playlist_{timestamp}.m3u8` as soon as Space is on-going (by changing `timestamp` in filename). Keep in mind, there are also some delays between the two so even if a Space ends, it takes some time for master_playlist.m3u8 to point to the "final" playlist_{timestamp}.m3u8 which includes the all the chunks (which is why I added a 1-minute delay in my patched version: to ensure I got all the chunks). You can see from the log above, even after the Space ended, the timestamp of playlist in master_playlist was still constantly changing. 

And *some* of these playlist_{timestamp}.m3u8 will just randomly be 404 for reasons I don't know, as demonstrated above. But it's on Twitter's end. My speculation is that when a Space just ended, Twitter's server still tries to create new playlist_xxx but since there is no new chunk coming, it failed somehow and did not create these playlists properly.

But this doesn't matter, because if we constantly retry after a Space ended, master m3u8 will eventually point to a valid, final playlist_{timestamp}.m3u8 which has all the chunks (and it no longer changes). In this case, it's `playlist_16760968411484182689.m3u8` (you can still get it from master_playlist.m3u8 even now). And we want to use this (regardless if others are 200 or not): because only this one has all the chunks. 


In summary, there is no point to even try to fetch or save any intermediate `playlist_{timestamp}.m3u8`, because it will only have a few chunks, if not straightly up 404. Since Twitter don't delete actual chunks or `master_playlist.m3u8` even for unarchived Space, all we need to do is to mark the dynamic URL or `master_playlist.m3u8` URL (which we already did) and use it to retrieve the final `playlist_{timestamp}.m3u8` once the Space ended.

**Comment by ef1500** [comment]
Sadly, due to the current state of Twitter's API, I will be postponing this feature addition until further notice.
We can work on a solution after we get this API business figured out! Any help is greatly appreciated.

**Comment by ef1500** [comment]
I believe that with the latest commit, this has been fixed. Please try it now and see if you still run into issues.
